### ISSUE TITLE
`invalid-argument-type` when passing overloaded function to `map`

### ISSUE THREAD

**Issue by auscompgeek** [astral-sh/ty#2546]
### Summary

On the following file foo.py:

```python
from typing import overload


@overload
def foo(x: int) -> str: ...

@overload
def foo(x: str) -> str: ...

def foo(x: int | str) -> str:
    return str(x)


def errors() -> None:
    for x in map(foo, range(1, 10)):
        print(x)
```

ty returns an error ([playground link](https://play.ty.dev/0d4b5ef1-33b2-455b-892d-17a6f09175e6)):

```console
> ty check foo.py
error[invalid-argument-type]: Argument to function `__new__` is incorrect
  --> foo.py:15:18
   |
14 | def errors() -> None:
15 |     for x in map(foo, range(1, 10)):
   |                  ^^^ Expected `(int | str, /) -> str | int`, found `Overload[(x: int) -> str, (x: str) -> str]`
16 |         print(x)
   |
info: Matching overload defined here
    --> stdlib/builtins.pyi:3866:13
     |
3864 |     else:
3865 |         @overload
3866 |         def __new__(cls, func: Callable[[_T1], _S], iterable: Iterable[_T1], /) -> Self: ...
     |             ^^^^^^^      ------------------------- Parameter declared here
3867 |         @overload
3868 |         def __new__(cls, func: Callable[[_T1, _T2], _S], iterable: Iterable[_T1], iter2: Iterable[_T2], /) -> Self: ...
     |
info: Non-matching overloads for function `__new__`:
info:   [_T1, _T2, Self](cls, func: (_T1, _T2, /) -> _S@map, iterable: Iterable[_T1], iter2: Iterable[_T2], /) -> Self
info:   [_T1, _T2, _T3, Self](cls, func: (_T1, _T2, _T3, /) -> _S@map, iterable: Iterable[_T1], iter2: Iterable[_T2], iter3: Iterable[_T3], /) -> Self
info:   [_T1, _T2, _T3, _T4, Self](cls, func: (_T1, _T2, _T3, _T4, /) -> _S@map, iterable: Iterable[_T1], iter2: Iterable[_T2], iter3: Iterable[_T3], iter4: Iterable[_T4], /) -> Self
info:   [_T1, _T2, _T3, _T4, _T5, Self](cls, func: (_T1, _T2, _T3, _T4, _T5, /) -> _S@map, iterable: Iterable[_T1], iter2: Iterable[_T2], iter3: Iterable[_T3], iter4: Iterable[_T4], iter5: Iterable[_T5], /) -> Self
info:   [Self](cls, func: (...) -> _S@map, iterable: Iterable[Any], iter2: Iterable[Any], iter3: Iterable[Any], iter4: Iterable[Any], iter5: Iterable[Any], iter6: Iterable[Any], /, *iterables: Iterable[Any]) -> Self
```

This results in an even more obscure error message when a class constructor is overloaded ([playground link](https://play.ty.dev/eebccddd-e66d-4f07-958a-cbc40b438037)), like:

```python
class Foo:
    @overload
    def __init__(self, x: int): ...
    @overload
    def __init__(self, x: str): ...
    def __init__(self, x) -> None:
        self.x = x
```

```console
> ty check foo.py
error[invalid-argument-type]: Argument to function `__new__` is incorrect
  --> foo.py:14:18
   |
13 | def errors() -> None:
14 |     for y in map(Foo, range(1, 10)):
   |                  ^^^ Expected `(int | str, /) -> Foo`, found `<class 'Foo'>`
15 |         print(y.x)
   |
info: Matching overload defined here
    --> stdlib/builtins.pyi:3866:13
[snip]
```

Possibly related issues, albeit closed:

- #822
- #771

### Version

ty 0.0.12

**Comment by dhruvmanila** [comment]
This seems specific to overloads because if I remove them then the error disappears.

Looking at the error message, I think it might be because the `func: Callable[[_T1], _S]` gets specialized into a single callable instead of capturing the overloads and so when the argument types are checked, there's a mismatch.

**Comment by carljm** [comment]
Simpler repro that doesn't involve `map` but has the same error message:

```py
from typing import overload, Callable

@overload
def foo(x: int) -> str: ...

@overload
def foo(x: str) -> str: ...

def foo(x: int | str) -> str:
    return str(x)

def f(x: Callable[[int | str], int | str]): ...

# error: [invalid-argument-type] "Argument to function `f` is incorrect: Expected `(int | str, /) -> int | str`, found `Overload[(x: int) -> str, (x: str) -> str]`"
f(foo)
```
https://play.ty.dev/4cf86577-5f55-4e17-b9ec-24d8cb3f87ec

So it doesn't seem like this even requires generics or paramspec -- it looks like a simpler issue in our callable assignability logic when overloads are involved? It looks to me like the overloaded `foo` should be a subtype of `Callable[[int | str], int | str]`, but we don't consider it one.

**Comment by dcreager** [comment]
It looks like the underlying issue is that we're intersecting the parameters of `foo` when seeing if it's assignable to the `Callable`. A similar example with `str` instead of `foo` [[playground](https://play.ty.dev/eca29e23-7847-4ac2-9845-bdb54de59421)]:

```py
def f[T, U](c: Callable[[T], U]) -> Callable[[T], U]:
    return c

# revealed: Overload[(object: object = "") -> str, (object: Buffer, encoding: str = "utf-8", errors: str = "strict") -> str]
reveal_type(into_callable(str))

# revealed: (Buffer, /) -> str
reveal_type(f(str))
```

**Comment by sharkdp** [comment]
Just adding a note here that this leads to some false positives on the typing conformance suite. Specifically the three diagnostics in `class DC2` [here](https://shark.fish/typing-conformance-report/dataclasses_transform_converter.html#L98). Found this while working on dataclass-transform `converter`s, trying to get these conformance tests to pass.

**Comment by Hugo-Polloli** [comment]
Hi! Thanks for all the pointers in this thread they really help, I'd like to work on this issue if no one else is actively working on it?

**Comment by carljm** [comment]
@Hugo-Polloli Sure!
